From 6235dd8200c52f81c55882bcaf5553715d9565cf Mon Sep 17 00:00:00 2001
From: Kyle Brenneman <kbrenneman@nvidia.com>
Date: Wed, 9 Sep 2026 12:31:14 -0600
Subject: [PATCH] wayland: Fix leaking buffer release events with explicit sync

If we're using explicit sync, then don't create a persistent event
queue for each swapchain.

We need a per-swapchain event queue for implicit or CPU sync, because we
have to rely on receiving wl_buffer::release events. With explicit sync,
we don't need those events.

If the server sends release events (which it's allowed but required to),
then the events will pile up the queue, since we never dispatch it.

Instead, only create a temporary queue long enough to receive the
zwp_linux_buffer_params_v1 events, and then set the resulting wl_buffer
back to the default queue.
---
 src/wayland/wayland-swapchain.c | 59 ++++++++++++++++++++++++++-------
 1 file changed, 47 insertions(+), 12 deletions(-)

diff --git a/src/wayland/wayland-swapchain.c b/src/wayland/wayland-swapchain.c
index 134fd08..b04c4d5 100644
--- a/src/wayland/wayland-swapchain.c
+++ b/src/wayland/wayland-swapchain.c
@@ -128,9 +128,27 @@ static struct wl_buffer *ShareDmaBuf(WlDisplayInstance *inst,
         uint32_t stride, uint32_t offset, uint32_t fourcc, uint64_t modifier)
 {
     DmaBufParamsCreateState state = {};
+    struct wl_event_queue *local_queue = NULL;
     struct zwp_linux_dmabuf_v1 *wrapper = NULL;
     struct zwp_linux_buffer_params_v1 *params = NULL;
 
+    if (queue == NULL)
+    {
+        /*
+         * When we're using explicit sync, we don't care about
+         * wl_buffer::release events, so we don't have a persistent event queue
+         * for the swapchain. Create a temporary one for the
+         * zwp_linux_buffer_params_v1 events.
+         */
+        local_queue = wl_display_create_queue(inst->wdpy);
+        if (local_queue == NULL)
+        {
+            goto done;
+        }
+
+        queue = local_queue;
+    }
+
     wrapper = wl_proxy_create_wrapper(inst->globals.dmabuf);
     if (wrapper == NULL)
     {
@@ -169,6 +187,16 @@ static struct wl_buffer *ShareDmaBuf(WlDisplayInstance *inst,
     {
         wl_proxy_wrapper_destroy(wrapper);
     }
+    if (local_queue != NULL)
+    {
+        // Set the wl_buffer to the display's queue, so that we can destroy our
+        // temporary event queue.
+        if (state.buffer != NULL)
+        {
+            wl_proxy_set_queue((struct wl_proxy *) state.buffer, NULL);
+        }
+        wl_event_queue_destroy(local_queue);
+    }
 
     return state.buffer;
 }
@@ -336,19 +364,26 @@ WlSwapChain *eplWlSwapChainCreate(WlDisplayInstance *inst, struct wl_surface *ws
     swapchain->present_fourcc = present_fourcc;
     swapchain->modifier = DRM_FORMAT_MOD_INVALID;
     swapchain->prime = prime;
-    if (inst->platform->priv->wl.display_create_queue_with_name != NULL)
+    if (inst->globals.syncobj == NULL)
     {
-        char name[64];
-        snprintf(name, sizeof(name), "EGLSurface(%u/%p)", wl_proxy_get_id((struct wl_proxy *) wsurf), swapchain);
-        swapchain->queue = inst->platform->priv->wl.display_create_queue_with_name(inst->wdpy, name);
-    }
-    else
-    {
-        swapchain->queue = wl_display_create_queue(inst->wdpy);
-    }
-    if (swapchain->queue == NULL)
-    {
-        goto done;
+        /*
+         * If we don't have explicit sync, then we'll need an event queue to
+         * receive wl_buffer::release events.
+         */
+        if (inst->platform->priv->wl.display_create_queue_with_name != NULL)
+        {
+            char name[64];
+            snprintf(name, sizeof(name), "EGLSurface(%u/%p)", wl_proxy_get_id((struct wl_proxy *) wsurf), swapchain);
+            swapchain->queue = inst->platform->priv->wl.display_create_queue_with_name(inst->wdpy, name);
+        }
+        else
+        {
+            swapchain->queue = wl_display_create_queue(inst->wdpy);
+        }
+        if (swapchain->queue == NULL)
+        {
+            goto done;
+        }
     }
 
     /*
